### ISSUE TITLE
Duplicate groups in the admin console of Keycloak

### ISSUE THREAD

**Issue by npriebe** [keycloak/keycloak#20342]
### Description

Hi,

It would be handy if it was possible to copy a group into Keycloak with all its attributes, child groups, members and assigned roles.
![Screenshot from 2023-05-15 08-21-14](https://github.com/keycloak/keycloak/assets/54103655/a6842364-0ed5-4fc2-95a3-ab83b6bb989c)
The duplicate option could then be added in the context menu under "Create child group".

### Motivation

The motivation for this is that you do not always need to create and assign a new group with many assigned roles, for example, but could simply copy and edit an existing one.
Or if the group hierarchy always has a basic configuration with assigned roles and users. These do not have to be added each time, but a group can simply be copied and then extended with the basic config.

**Comment by ReCodee** [comment]
Hi, I'd like to work on this feature. 
@ssilvert @jonkoops Is this issue ticket relevant?

**Comment by edewit** [comment]
@ReCodee sorry for the late reply, yes this is something we are still interested in and we would love for this to be a feature. Just create a PR and ping me on it and we can work towards getting that in

**Comment by ssilvert** [comment]
@ReCodee @edewit We are likely to run into issues when we need to duplicate a very large tree of groups.  Some can have thousands of children.  Maybe we error out if there are too many children?

**Comment by ReCodee** [comment]
> @ReCodee @edewit We are likely to run into issues when we need to duplicate a very large tree of groups. Some can have thousands of children. Maybe we error out if there are too many children?

@ssilvert Thanks, I didn't think about the feature from this perspective. Are we gonna restrict the duplication if a group has too many children? We can even give some kind of warning/confirmation to the user if the duplicating group has more than usual amount of children.

**Comment by npriebe** [comment]
Thank you for your efforts. @ReCodee @ssilvert @xianli123 

Would it also make sense to have the option in the modal popup of duplicating to select what should be duplicated?
You could then offer checkboxes here.

Example:

[X] Attributes
[ ] Roles
[X] User
[X] Sub-Groups

If a checkbox is selected, it will also be duplicated.
With the subgroups, however, it can be a little difficult to decide which subgroups to duplicate for.
But maybe you have a better idea.

**Comment by ssilvert** [comment]
@npriebe I think the modal is a good idea.  Not sure what you mean about the difficulty with subgroups?  It would just be all or nothing, right?

@edewit has the most experience with the Groups section.  So he's the best technical consultant on it.

**Comment by npriebe** [comment]
@ssilvert I mean, it would be practical if you could still choose which properties of the groups should be duplicated (attributes, roles, users, etc.). But that would be difficult, because theoretically you would have to do that for each subgroup separately.
But I think that the modal is a good solution and that each property is duplicated afterwards.

**Comment by ssilvert** [comment]
I think the all-or-nothing approach is a good compromise.  Doing it for each subgroup would be really complicated.

**Comment by xianli123** [comment]
@ssilvert @npriebe Do you mean this way? 
<img width="457" alt="Screenshot 2023-06-29 at 14 33 17" src="https://github.com/keycloak/keycloak/assets/53931519/0c645566-b16d-4ec9-9d41-98fd4e0ef084">

**Comment by npriebe** [comment]
@xianli123 Exactly, but the problem is that you would theoretically have to decide on the settings for each sub-group individually. Or you would lack the information that the settings would affect all the subgroups.

**Comment by xianli123** [comment]
@npriebe That's really complicated. Each sub-group may have its own sub-groups. It's a very complex situation.

**Comment by npriebe** [comment]
@xianli123 Yes, that's why the simplest thing would be as @ssilvert  says, all-or-nothing.

**Comment by xianli123** [comment]
@npriebe It may look like this one. Users can duplicate all things in this group or don't duplicate the group. Right?

![image](https://github.com/keycloak/keycloak/assets/53931519/78f5521e-5e0c-4ee2-bd98-9d8e7ad4a1bf)

**Comment by npriebe** [comment]
@xianli123 Yes

**Comment by xianli123** [comment]
@npriebe thanks. That is a simple way for users. Even the description under the **Duplicate group** can be removed. It looks like the screenshot below.
<img width="478" alt="Screenshot 2023-06-29 at 15 11 41" src="https://github.com/keycloak/keycloak/assets/53931519/9ae8d437-6087-45b7-a1df-6185ff270a14">

**Comment by ReCodee** [comment]
@npriebe @xianli123 Thanks for the clarification

**Comment by drusso-uni-doc** [comment]
HI. Has this feature already been implemented in some version of keycloak? Thank you

**Comment by ssilvert** [comment]
> HI. Has this feature already been implemented in some version of keycloak? Thank you

No, but we probably should take a look at it when we can as it has gotten a few upvotes.

**Comment by mmiggi** [comment]
yea we woulde also look forward to use that feature, since we manage a few things with similar looking groups. When we make a new group we woulde like them to have the same Attributes. Now we have to add 20 Attributes everytime we make a new one.

**Comment by ssilvert** [comment]
> > > @ReCodee @edewit We are likely to run into issues when we need to duplicate a very large tree of groups. Some can have thousands of children. Maybe we error out if there are too many children?
> > 
> > 
> > @ssilvert Thanks, I didn't think about the feature from this perspective. Are we gonna restrict the duplication if a group has too many children? We can even give some kind of warning/confirmation to the user if the duplicating group has more than usual amount of children.
> 
> @ssilvert Could you advise on what would be considered too many children in the group, which would result in displaying an error to the user? Thx

@agagancarczyk I don't know.  It would take some experimentation to find out.  Do we have a way to determine the number of children before we run into trouble?
